### ISSUE TITLE
`autorun` is not working as expected in the recent version of flutter.

### ISSUE THREAD

**Issue by wiseaidev** [mobxjs/mobx.dart#897]
Hey everyone. I just wanted to give you an update on the previous issue. I found that the issue is reproducible on the current version of flutter, as shown in the following image:

![112](https://user-images.githubusercontent.com/62179149/215248843-3275f2c6-4474-4237-a660-e7fe09a23b5b.PNG)

However, if you try to change the lower bound of [the SDK versions](https://github.com/wiseaidev/flutter-mobx-autorun-issue/blob/main/pubspec.yaml#L23) to `2.16.8`, the function works as expected:

```diff
-   sdk: '>=2.18.6 <3.0.0'
+   sdk: '>=2.16.8 <3.0.0'
```

![111](https://user-images.githubusercontent.com/62179149/215249303-3eb10f3e-323b-4aa1-a6fb-1fd4a2d2b3f2.PNG)

It seems like the widget's behavior has changed in the newer flutter version. In particular, it appears that the constructor has been updated in a way that causes this issue.

- Widget constructors then:

```dart
  const MyApp({Key? key}) : super(key: key);
  const MySplashPage({Key? key, required this.title}) : super(key: key);
```

- Widget constructors now:

```dart
  const MyApp({super.key});
  const MySplashPage({super.key, required this.title});
```

Have a look at [the previous ticket](https://github.com/mobxjs/mobx.dart/issues/895) for more details on this issue.

**Comment by RossHS** [comment]
@wiseaidev Greetings, I met exactly the same problem. It seems to me that the reason lies in the special processing of the `InheritedWidget` (on which the `provider` relies) by the Flutter framework. 

so if you try to convert the widget to something like this (wrapping it with `Builder` widget), it should help

```dart
Builder(
      builder: (context) {
        final appState = Provider.of<Counter>(context);
        return ReactionBuilder(
          builder: (context) {
            return reaction(
              (_) => appState.currentPage,
              (v)  => print('new value: $v'),
            );
          },
          child: ...
        );
      }
    );
```

**Comment by amondnet** [comment]
@wiseaidev 
I tested with the version below.
1. flutter 3.3.8: works
2. flutter 3.3.10: not works
3. flutter 3.7.0: not works


If I change it like below, it works fine.

```dart
class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    final appState = context.read<AppState>();
    return MaterialApp(
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      debugShowCheckedModeBanner: false,
      home: ReactionBuilder(
        builder: (context) {
          return reaction(
            (_) => appState.currentPage,
            (v) {
              print("A reaction has been triggered! $v");
            },
          );
        },
        child: Observer(
          name: "CurrentScreen",
          builder: (context) {
            switch (context.read<AppState>().currentPage) {
              case AppPage.splash:
                return const MySplashPage(title: "Splash Page");
              case AppPage.home:
                return const MySplashPage(title: "Home Page");
              case AppPage.login:
                return const Placeholder();
            }
          },
        ),
      ),
    );
  }
}
```

**Comment by amondnet** [comment]
```dart
import 'package:flutter/material.dart';
import 'package:flutter_mobx/flutter_mobx.dart';
import 'package:mobx/mobx.dart';
import 'package:mobx_autorun_issue/store.dart';
import 'package:provider/provider.dart';

void main() async {
  WidgetsFlutterBinding.ensureInitialized();
  runApp(
    Provider(
      create: (_) => AppState()..initialize(),
      child: const MyApp(),
    ),
  );
}
var reactionTriggered = false;

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    // final appState = context.read<AppState>();
    return MaterialApp(
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      debugShowCheckedModeBanner: false,
      home: ReactionBuilder(
        builder: (context) {
          return reaction(
            (_) => context.read<AppState>().currentPage,
            (v) {
              print("A reaction has been triggered! $v");
              reactionTriggered = true;
            },
          );
        },
        child: Observer(
          name: "CurrentScreen",
          builder: (context) {
            switch (context.read<AppState>().currentPage) {
              case AppPage.splash:
                return const MySplashPage(title: "Splash Page");
              case AppPage.home:
                return const MySplashPage(title: "Home Page");
              case AppPage.login:
                return const Placeholder();
            }
          },
        ),
      ),
    );
  }
}

class MySplashPage extends StatefulWidget {
  const MySplashPage({super.key, required this.title});

  final String title;

  @override
  State<MySplashPage> createState() => _MySpashPageState();
}

class _MySpashPageState extends State<MySplashPage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () {
          print("Am action has been fired!");
          context.read<AppState>().goTo(
                AppPage.home,
              );
        },
        tooltip: 'Smash It.',
        child: const Icon(Icons.app_registration_outlined),
      ),
    );
  }
}

```

```dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility in the flutter_test package. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:mobx_autorun_issue/main.dart';
import 'package:mobx_autorun_issue/store.dart';
import 'package:provider/provider.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(TestApp());

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.app_registration_outlined));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(reactionTriggered, true);
  });
}

class TestApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Provider(
      create: (_) => AppState()..initialize(),
      child: const MyApp(),
    );
  }
}

```

**Comment by RossHS** [comment]
@amondnet I'm sorry, I really don't understand what's wrong here 😂, but here are two code examples + video, in one case the reaction works, in the other it doesn't.

mobx store class - 

```dart
import 'package:mobx/mobx.dart';

part 'reaction_builder_mobx.g.dart';

class Counter = CounterBase with _$Counter;

abstract class CounterBase with Store {
  @observable
  var value = 0;

  @action
  void increment() {
    value++;
  }

  @action
  void decrement() {
    value--;
  }
}
```

first code example (doesn`t work)
```dart
    return ReactionBuilder(
      builder: (context) => reaction(
        (_) => Provider.of<Counter>(context).value,
        (res) {
          log.d('current counter value - $res');
          if (res != 10) return;
          ScaffoldMessenger.of(context).showSnackBar(SnackBar(
            content: Text('current counter val - $res == 10!!!'),
          ));
        },
      ),
      child: .....
```

and the second one (everything is the same except for access to the provider)
```dart
    final counter = Provider.of<Counter>(context);
    return ReactionBuilder(
      builder: (context) => reaction(
        (_) => counter.value,
        (res) {
          log.d('current counter value - $res');
          if (res != 10) return;
          ScaffoldMessenger.of(context).showSnackBar(SnackBar(
            content: Text('current counter val - $res == 10!!!'),
          ));
        },
      ),
      child: ...
```


https://user-images.githubusercontent.com/30704653/215399112-47d79fbc-144a-450a-87a5-d0c5a02aec34.mp4

**Comment by amondnet** [comment]
@RossHS 
Provider.of<Counter>(context).value =>
 Provider.of<Counter>(context, **listen: false**).value

https://github.com/rrousselGit/provider#reading-a-value

```dart
    return ReactionBuilder(
      builder: (context) => reaction(
        (_) => Provider.of<Counter>(context, listen: false).value,
        (res) {
          log.d('current counter value - $res');
          if (res != 10) return;
          ScaffoldMessenger.of(context).showSnackBar(SnackBar(
            content: Text('current counter val - $res == 10!!!'),
          ));
        },
      ),
      child: .....
```

**Comment by RossHS** [comment]
@amondnet 
You are right, thank you very much. All because of my inattention 🌚

**Comment by wiseaidev** [comment]
Hey folks,

Thanks for your contributions to the conversation. I tried both of your suggestions in the app I am currently building, but, unfortunately, it is still not working.

![213124](https://user-images.githubusercontent.com/62179149/215465919-1dee491d-2d70-45e4-8437-0140434e1509.PNG)

> The reaction occurs only when the observable's value is a different value from the previous one.

Exactly. But the thing is that it is not triggered for the first time. I think the first image I shared is somewhat misleading. However, I am not sure what is wrong, Mobx? Provider? Flutter? Gradle? Android studio? I literally have no idea. However, for now, I am going to downgrade flutter to the version that was working before.

I think that's enough flutter for the week cause this issue is so trippy. I will keep you updated on this matter. Have fun!
